pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -137,9 +159,10 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
         processed_batch_end_outputs = self._prepare_outputs(batch_end_outputs, batch_mode=True)
 
         # hook
-        self.trainer.call_hook(
-            "on_train_batch_end", processed_batch_end_outputs, batch, self.batch_idx, self._dataloader_idx
-        )
+        if not isinstance(self.batch_loop, IteratorBatchProcessor):
+            self.trainer.call_hook(
+                "on_train_batch_end", processed_batch_end_outputs, batch, self.batch_idx, self._dataloader_idx
+            )
         self.trainer.call_hook("on_batch_end")
         self.trainer.logger_connector.on_batch_end()
 
```

Comment:
do these also need to be under the check?

Reply:
We can still call this since the contract says that the `training_step` should only optimize the model with one batch (although it can kickoff async preprocessing of future batches):

```
The ``training_step`` should only optimize the model with one batch for the sake of API and
reporting consistency (TODO: consider removing this limitation).
```

We may want to remove this constraint from the contract in future though. We'll have to reconsider this then.